testing/integration/upgrade_broken_package_test.go

```diff
@@ -0,0 +1,92 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+//go:build integration
+
+package integration
+
+import (
+	"context"
+	"fmt"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	atesting "github.com/elastic/elastic-agent/pkg/testing"
+	"github.com/elastic/elastic-agent/pkg/testing/define"
+	"github.com/elastic/elastic-agent/testing/upgradetest"
+	agtversion "github.com/elastic/elastic-agent/version"
+)
+
+func TestUpgradeBrokenPackageVersion(t *testing.T) {
+	define.Require(t, define.Requirements{
+		Local: false, // requires Agent installation
+		Sudo:  true,  // requires Agent installation
+	})
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// Start at the build version as we want to test the retry
+	// logic that is in the build.
+	startFixture, err := define.NewFixture(t, define.Version())
+	require.NoError(t, err)
+
+	// Upgrade to an old build, see `BackwardTwoMinors` for why.
+	upgradeToVersion, err := upgradetest.BackwardTwoMinors(define.Version())
```

Comment:
Why do we use BackwardTwoMinors here? 
We already know that GetPreviousMinor() makes assumptions about the version numbers (set the patch version to 0, return a fixed version for 7.x), why not taking the version from artifact API as it was before the refactor?

Reply:
Almost all the tests where written completely different, but basically doing the same thing. I tried my best to merge them onto a path that is very similar and easy to follow. To me read ability is the most important thing, because if I cannot open the test and easily read and understand what is going on then its useless.

I tried my best to align the tests in a common pattern. If I have change the meaning of what the test is testing please let me know I will fix it.

Does this change the meaning of this test?

For why we use `BackwardTwoMinors` check the docstring on that function. It explains it.

Reply:
It doesn't change the meaning of the test, I am not that comfortable with us not using the latest patch for a given <major>.<minor>

Reply:
Are you saying that you want it to only go down a single version? I could possibly switch all to that way, and then possibly only go down 2 in the case the last 1 doesn't exist.

Reply:
Note: this is not blocking for the merge

It's not the number of version we downgrade to: in this specific case it doesn't really matter how far back we go, we just want to test that even with a broken package version we can upgrade to something else.

The thing making me uncomfortable is that from `8.11.0-SNAPSHOT` we are going back to `8.10.0` or `8.9.0` instead of `8.10.3` or `8.9.2` because we cannot guess how patch versions a certain major.minor combination has.

As a personal preference I favored using [artifact api version list](https://artifacts-api.elastic.co/v1/versions) to search for a previous version instead of computing a likely value but it's true that unless we have serious bugs on our `x.y.0` versions this won't break any time soon for 8.x

7.x may break when 7.17.10 will no longer be available through artifact api (have a look at GetPreviousMinor() code)
9.x if and when it will land will need also some special handling in GetPreviousMinor()

As long as we use this to go a couple of minors  back from the latest 8.x we will not fall into those cases so it comes down to preference.

Reply:
I have updated the code to take the same idea you had from your other PR and use it in this PR. Now it uses the artifacts API to find the previous version that has a snapshot build and uses that for the test.